internal/controlplane/handlers_projects.go

```diff
@@ -154,83 +154,87 @@ func (s *Server) getImmediateChildrenProjects(ctx context.Context, projectID uui
+// or creates a sub-project if permitted (if req.Context.Project is set).
+// The project name must be unique within the parent project, or across all projects if
+// creating a top-level project.
+//
+// Because this may be called in a non-project context (to create a top-level project),
+// we need to _explicitly_ perform project authorization checks if project is non-nil/empty.
 func (s *Server) CreateProject(
 	ctx context.Context,
 	req *minderv1.CreateProjectRequest,
 ) (*minderv1.CreateProjectResponse, error) {
-	entityCtx := engcontext.EntityFromContext(ctx)
-	projectID := entityCtx.Project.ID
-
-	if !features.ProjectAllowsProjectHierarchyOperations(ctx, s.store, projectID) {
-		return nil, util.UserVisibleError(codes.PermissionDenied,
-			"project does not allow project hierarchy operations")
+	parentProjectID, err := getProjectIDFromRequest(req)
+	if err != nil && !errors.Is(err, ErrNoProjectInContext) {
+		return nil, err
 	}
 
-	tx, err := s.store.BeginTransaction()
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "error starting transaction: %v", err)
-	}
-	defer s.store.Rollback(tx)
+	var project *db.Project
+	if parentProjectID != uuid.Nil {
+		// Verify permissions if we have a parent
+		relationName := relationAsName(minderv1.Relation_RELATION_CREATE)
+		s.authzClient.Check(ctx, relationName, parentProjectID)
 
-	qtx := s.store.GetQuerierWithTransaction(tx)
+		if !features.ProjectAllowsProjectHierarchyOperations(ctx, s.store, parentProjectID) {
+			return nil, util.UserVisibleError(codes.PermissionDenied,
+				"project does not allow project hierarchy operations")
+		}
+		tx, err := s.store.BeginTransaction()
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "error starting transaction: %v", err)
+		}
+		defer s.store.Rollback(tx)
+		qtx := s.store.GetQuerierWithTransaction(tx)
 
-	parent, err := qtx.GetProjectByID(ctx, projectID)
-	if err != nil {
-		if errors.Is(err, sql.ErrNoRows) {
-			return nil, util.UserVisibleError(codes.NotFound, "project not found")
+		project, err = s.projectCreator.ProvisionChildProject(ctx, qtx, parentProjectID, req.Name)
+		if err != nil {
+			return nil, err
 		}
-		return nil, status.Errorf(codes.Internal, "error getting project: %v", err)
-	}
 
-	// TODO: Remove this once we handle a full hierarchy
-	if parent.ParentID.Valid {
-		return nil, util.UserVisibleError(codes.InvalidArgument, "cannot create subproject of a subproject")
-	}
+		if err := s.store.Commit(tx); err != nil {
+			return nil, status.Errorf(codes.Internal, "error committing transaction: %v", err)
+		}
 
-	if err := projects.ValidateName(req.Name); err != nil {
-		return nil, util.UserVisibleError(codes.InvalidArgument, "invalid project name: %v", err)
-	}
+	} else {
+		// This is a top-level project creation request.
+		// We need to check if the user has the right to create projects in the system.
+		if !flags.Bool(ctx, s.featureFlags, flags.ProjectCreateDelete) {
```

Comment:
Do we need a feature flag for this?

Reply:
It feels like good practice, even if I'm going to turn it on shortly.